### PR TITLE
BOM: Feature complete SPDX and license packages 

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR contains the final updates to the go packages needed to produce the SPDX Bill of Materials for Kubernetes releases. This code now produces SPDX compliant manifests from naked files, docker archives, and images pulled from registries.  

 This PR does the following:

- Rips the SPDX code from the license package code into a new spdx pkg
- Moves the license lookup code into a new `license.Catalog` object 
- Implements the first set of image analyzers to enrich the SPDX data for the distroless and go-runner base images
- Corrects a bug with the license downloader where license data was not available when first downloaded.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/release/issues/1837

#### Special notes for your reviewer:

This is the first commit of the final form of the packages. The packages are not in use yet by anything. The tests and fakes for the spdx package are coming in a follow-up

#### Does this PR introduce a user-facing change?
```release-note
- New `SPDX` package for generating SPDX compliant manifests of artifacts.
- The `license` package now includes a new `Catalog` object to interact with spdx license data 
- First set of image analyzers to enrich the BOM generated for the go-runner and distroless base images
- Corrects a bug with the license downloader where license data was not available when first downloaded. 
```

### CODE REVIEW

File: pkg/license/catalog.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+func (catalog *Catalog) WriteLicensesAsText(targetDir string) error {
+	logrus.Info("Writing SPDX licenses to " + targetDir)
+	if catalog.List.Licenses == nil {
+		return errors.New("unable to write licenses, they have not been loaded yet")
+	}
+	for _, l := range catalog.List.Licenses {
+		if err := l.WriteText(filepath.Join(targetDir, l.LicenseID+".txt")); err != nil {
+			return errors.Wrapf(err, "while writing license %s", l.LicenseID)
+		}
+	}
```
Comment (saschagrunert) [inline pkg/license/catalog.go:88]: We could think about making this parallel.

File: pkg/license/catalog.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	// Create the license downloader
+	doptions := DefaultDownloaderOpts
+	doptions.CacheDir = opts.CacheDir
+	downloader, err := NewDownloaderWithOptions(doptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating downloader")
+	}
+	catalog = &Catalog{
+		Downloader: downloader,
+		Options:    DefaultCatalogOpts,
```
[the remark this answers is not in the document]
Comment (puerco) [reply pkg/license/catalog.go:51]: Nice catch! thanks :)

File: pkg/license/catalog.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+// CatalogOptions are the spdx settings
+type CatalogOptions struct {
+	CacheDir string
+}
+
+// Validate checks the spdx options
+func (o *CatalogOptions) Validate() error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (puerco) [reply pkg/license/catalog.go]: I've stripped the validation entirely for now

File: pkg/license/catalog.go (hunk 1 above)
Comment (puerco) [reply pkg/license/catalog.go]: Ready, rewritten to save the files to disk in parallel.

**Review by saschagrunert — APPROVED** [review]

**Comment by saschagrunert** [comment]
This PR is heavy and I did not manage to review in every detail. Let's iterate on it and get the main functionality in. Great work! :partying_face:

**Pull request merged**